File: Testing/tests/client/Tests/BasicCSharp/TestTryCatchFinallyBlocks.cs
using Bridge;
using Bridge.QUnit;
using System;

#pragma warning disable 162	// CS0162: Unreachable code detected. Disable because we want to assert that code does not reach unreachable parts

namespace ClientTestLibrary
{
    [FileName("testTryCatchFinallyBlocks.js")]
    internal class Data
    {
        public int Count
        {
            get;
            set;
        }
    }

    // Tests try and catch blocks
    internal class TestTryCatchFinallyBlocks
    {
        #region Tests

        public static void SimpleTryCatchFinally(Assert assert)
        {
            assert.Expect(1);

            var data = new Data();
            TryCatchFinally(data);

            assert.Equal(data.Count, 2, "TryCatchFinally() executes");
        }

        public static void CaughtExceptions(Assert assert)
        {
            assert.Expect(4);

            var data = new Data();
            TryCatchFinallyWithCaughtException(data);

            assert.Equal(data.Count, 7, "Exception catch, Finally works");

            data = new Data();
            TryCatchFinallyWithCaughtTypedException(data);

            assert.Equal(data.Count, 7, "Typed exception catch, Finally works");

            data = new Data();
            var exceptionMessage = TryCatchFinallyWithCaughtArgumentException(data);

            assert.Equal(exceptionMessage, "catch me", "Typed exception catch with exception message");
            assert.Equal(data.Count, 7, "Typed exception catch with exception message, Finally works");
        }

        public static void ThrownExceptions(Assert assert)
        {
            assert.Expect(16);

            //#230
            assert.Throws(TryCatchFinallyWithNotCaughtTypedException, "catch me", "A. Typed exception is not caught");
            assert.Ok(IsATry, "A. exception not caught - try section called");
            assert.Ok(!IsACatch, "A. exception not caught - catch section not called");
            assert.Ok(IsAFinally, "A. exception not caught - finally section called");

            //#229
            assert.Throws(TryCatchWithNotCaughtTypedExceptionAndArgument, "catch me", "[#229] B. Typed exception is not caught; and argument");
            assert.Ok(IsBTry, "B. exception not caught - try section called");
            assert.Ok(!IsBCatch, "B. exception not caught - catch section not called");
            assert.Ok(IsBFinally, "B. exception not caught - finally section called");

            //#231
            assert.Throws(TryCatchWithRethrow, "catch me", "[#231] C. Rethrow");
            assert.Ok(IsCTry, "C. exception caught and re-thrown  - try section called");
            assert.Ok(IsCCatch, "C. exception caught and re-thrown  - catch section called");
            assert.Ok(IsCFinally, "C. exception caught and re-thrown  - finally section called");

            assert.Throws(TryCatchWithRethrowEx, new Func<object, bool>((error) =>
            {
                return error.ToString() == "catch me";
            }), "D. Rethrow with parameter");
            assert.Ok(IsDTry, "D. exception caught and re-thrown  - try section called");
            assert.Ok(IsDCatch, "D. exception caught and re-thrown  - catch section called");
            assert.Ok(IsDFinally, "D. exception caught and re-thrown  - finally section called");
        }

        #endregion Tests

        private static void TryCatchFinally(Data data)
        {
            try
            {
                data.Count++;
            }
            catch
            {
            }
            finally
            {
                data.Count++;
            }
        }

        #region CaughtExceptions

        private static void TryCatchFinallyWithCaughtException(Data data)
        {
            try
            {
                data.Count = data.Count + 1;
                throw new Exception();
                data.Count = data.Count - 1;
            }
            catch
            {
                data.Count = data.Count + 2;
            }
            finally
            {
                data.Count = data.Count + 4;
            }
        }

        private static void TryCatchFinallyWithCaughtTypedException(Data data)
        {
            try
            {
                data.Count = data.Count + 1;
                throw new Exception("catch me");
                data.Count = data.Count - 1;
            }
            catch (Exception)
            {
                data.Count = data.Count + 2;
            }
            finally
            {
                data.Count = data.Count + 4;
            }
        }

        private static string TryCatchFinallyWithCaughtArgumentException(Data data)
        {
            try
            {
                data.Count = data.Count + 1;
                throw new ArgumentException("catch me");
                data.Count = data.Count - 1;
            }
            catch (ArgumentException ex)
            {
                data.Count = data.Count + 2;

                return ex.Message;
            }
            finally
            {
                data.Count = data.Count + 4;
            }
        }

        #endregion CaughtExceptions

        #region ThrownExceptions

        public static bool IsATry
        {
            get;
            set;
        }

        public static bool IsACatch
        {
            get;
            set;
        }

        public static bool IsAFinally
        {
            get;
            set;
        }

        private static void TryCatchFinallyWithNotCaughtTypedException()
        {
            IsATry = false;
            IsACatch = false;
            IsAFinally = false;

            try
            {
                IsATry = true;
                throw new Exception("catch me");
                IsATry = false;
            }
            catch (ArgumentException)
            {
                IsACatch = true;
            }
            finally
            {
                IsAFinally = true;
            }
        }

        public static bool IsBTry
        {
            get;
            set;
        }

        public static bool IsBCatch
        {
            get;
            set;
        }

        public static bool IsBFinally
        {
            get;
            set;
        }

        private static void TryCatchWithNotCaughtTypedExceptionAndArgument()
        {
            IsBTry = false;
            IsBCatch = false;
            IsBFinally = false;

            try
            {
                IsBTry = true;
                throw new Exception("catch me");
                IsBTry = false;
            }
            catch (InvalidCastException ex)
            {
                var s = ex.Message;
                IsBCatch = true;
            }
            finally
            {
                IsBFinally = true;
            }
        }

        public static bool IsCTry
        {
            get;
            set;
        }

        public static bool IsCCatch
        {
            get;
            set;
        }

        public static bool IsCFinally
        {
            get;
            set;
        }

        private static void TryCatchWithRethrow()
        {
            IsCTry = false;
            IsCCatch = false;
            IsCFinally = false;

            try
            {
                IsCTry = true;
                throw new InvalidOperationException("catch me");
                IsCTry = false;
            }
            catch (Exception)
            {
                IsCCatch = true;
                throw;
            }
            finally
            {
                IsCFinally = true;
            }
        }

        public static bool IsDTry
        {
            get;
            set;
        }

        public static bool IsDCatch
        {
            get;
            set;
        }

        public static bool IsDFinally
        {
            get;
            set;
        }

        private static void TryCatchWithRethrowEx()
        {
            IsDTry = false;
            IsDCatch = false;
            IsDFinally = false;

            try
            {
                IsDTry = true;
                throw new ArgumentException("catch me");
                IsDTry = false;
            }
            catch (Exception ex)
            {
                IsDCatch = true;
                throw ex;
            }
            finally
            {
                IsDFinally = true;
            }
        }

        #endregion ThrownExceptions
    }
}

File: Testing/www/js/testTryCatchFinallyBlocks.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.Data', {
    config: {
        properties: {
            Count: 0
        }
    }
});

Bridge.define('ClientTestLibrary.TestTryCatchFinallyBlocks', {
    statics: {
        config: {
            properties: {
                IsATry: false,
                IsACatch: false,
                IsAFinally: false,
                IsBTry: false,
                IsBCatch: false,
                IsBFinally: false,
                IsCTry: false,
                IsCCatch: false,
                IsCFinally: false,
                IsDTry: false,
                IsDCatch: false,
                IsDFinally: false
            }
        },
        simpleTryCatchFinally: function (assert) {
            assert.expect(1);

            var data = new ClientTestLibrary.Data();
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchFinally(data);

            assert.equal(data.getCount(), 2, "TryCatchFinally() executes");
        },
        caughtExceptions: function (assert) {
            assert.expect(4);

            var data = new ClientTestLibrary.Data();
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchFinallyWithCaughtException(data);

            assert.equal(data.getCount(), 7, "Exception catch, Finally works");

            data = new ClientTestLibrary.Data();
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchFinallyWithCaughtTypedException(data);

            assert.equal(data.getCount(), 7, "Typed exception catch, Finally works");

            data = new ClientTestLibrary.Data();
            var exceptionMessage = Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchFinallyWithCaughtArgumentException(data);

            assert.equal(exceptionMessage, "catch me", "Typed exception catch with exception message");
            assert.equal(data.getCount(), 7, "Typed exception catch with exception message, Finally works");
        },
        thrownExceptions: function (assert) {
            assert.expect(16);

            //#230
            assert.throws(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchFinallyWithNotCaughtTypedException, "catch me", "A. Typed exception is not caught");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsATry(), "A. exception not caught - try section called");
            assert.ok(!Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsACatch(), "A. exception not caught - catch section not called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsAFinally(), "A. exception not caught - finally section called");

            //#229
            assert.throws(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchWithNotCaughtTypedExceptionAndArgument, "catch me", "[#229] B. Typed exception is not caught; and argument");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsBTry(), "B. exception not caught - try section called");
            assert.ok(!Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsBCatch(), "B. exception not caught - catch section not called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsBFinally(), "B. exception not caught - finally section called");

            //#231
            assert.throws(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchWithRethrow, "catch me", "[#231] C. Rethrow");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsCTry(), "C. exception caught and re-thrown  - try section called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsCCatch(), "C. exception caught and re-thrown  - catch section called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsCFinally(), "C. exception caught and re-thrown  - finally section called");

            assert.throws(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).tryCatchWithRethrowEx, function (error) {
                return error.toString() === "catch me";
            }, "D. Rethrow with parameter");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsDTry(), "D. exception caught and re-thrown  - try section called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsDCatch(), "D. exception caught and re-thrown  - catch section called");
            assert.ok(Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).getIsDFinally(), "D. exception caught and re-thrown  - finally section called");
        },
        tryCatchFinally: function (data) {
            try {
                data.setCount(data.getCount()+1);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
            }
            finally {
                data.setCount(data.getCount()+1);
            }
        },
        tryCatchFinallyWithCaughtException: function (data) {
            try {
                data.setCount(data.getCount() + 1);
                throw new Bridge.Exception();
                data.setCount(data.getCount() - 1);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                data.setCount(data.getCount() + 2);
            }
            finally {
                data.setCount(data.getCount() + 4);
            }
        },
        tryCatchFinallyWithCaughtTypedException: function (data) {
            try {
                data.setCount(data.getCount() + 1);
                throw new Bridge.Exception("catch me");
                data.setCount(data.getCount() - 1);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                data.setCount(data.getCount() + 2);
            }
            finally {
                data.setCount(data.getCount() + 4);
            }
        },
        tryCatchFinallyWithCaughtArgumentException: function (data) {
            try {
                data.setCount(data.getCount() + 1);
                throw new Bridge.ArgumentException("catch me");
                data.setCount(data.getCount() - 1);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                var ex;
                if (Bridge.is($e, Bridge.ArgumentException)) {
                    ex = $e;
                    data.setCount(data.getCount() + 2);

                    return ex.getMessage();
                }
                else {
                    throw $e;
                }
            }
            finally {
                data.setCount(data.getCount() + 4);
            }
        },
        tryCatchFinallyWithNotCaughtTypedException: function () {
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsATry(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsACatch(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsAFinally(false);

            try {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsATry(true);
                throw new Bridge.Exception("catch me");
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsATry(false);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                if (Bridge.is($e, Bridge.ArgumentException)) {
                    Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsACatch(true);
                }
                else {
                    throw $e;
                }
            }
            finally {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsAFinally(true);
            }
        },
        tryCatchWithNotCaughtTypedExceptionAndArgument: function () {
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBTry(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBCatch(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBFinally(false);

            try {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBTry(true);
                throw new Bridge.Exception("catch me");
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBTry(false);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                var ex;
                if (Bridge.is($e, Bridge.InvalidCastException)) {
                    ex = $e;
                    var s = ex.getMessage();
                    Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBCatch(true);
                }
                else {
                    throw $e;
                }
            }
            finally {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsBFinally(true);
            }
        },
        tryCatchWithRethrow: function () {
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCTry(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCCatch(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCFinally(false);

            try {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCTry(true);
                throw new Bridge.InvalidOperationException("catch me");
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCTry(false);
            }
            catch ($e) {
                $e = Bridge.Exception.create($e);
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCCatch(true);
                throw $e;
            }
            finally {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsCFinally(true);
            }
        },
        tryCatchWithRethrowEx: function () {
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDTry(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDCatch(false);
            Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDFinally(false);

            try {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDTry(true);
                throw new Bridge.ArgumentException("catch me");
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDTry(false);
            }
            catch (ex) {
                ex = Bridge.Exception.create(ex);
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDCatch(true);
                throw ex;
            }
            finally {
                Bridge.get(ClientTestLibrary.TestTryCatchFinallyBlocks).setIsDFinally(true);
            }
        }
    }
});



Bridge.init();